tests/test_churn.py

```diff
@@ -0,0 +1,389 @@
+    spark = (SparkSession
+             .builder
+             .appName("churn_test")
+             .getOrCreate())
+
+    # teardown
+    request.addfinalizer(lambda: spark.stop())
+
+    return spark
+
+
+"""
+Calendar for reference
+
+    January 2017
+Su Mo Tu We Th Fr Sa
+ 1  2  3  4  5  6  7
+ 8  9 10 11 12 13 14
+15 16 17 18 19 20 21
+22 23 24 25 26 27 28
+29 30 31
+
+# General guidelines
+
+There are a few gotches that you should look out for while creating
+new tests for this suite that are all churn specific.
+
+* assign rows new client_ids if you want them to show up as unique
+rows in the output datafrane
+
+"""
+
+main_schema = StructType([
+    StructField("app_version",           StringType(), True),
+    StructField("attribution",           StructType([
+        StructField("source",            StringType(), True),
+        StructField("medium",            StringType(), True),
+        StructField("campaign",          StringType(), True),
+        StructField("content",           StringType(), True)]), True),
+    StructField("channel",               StringType(),  True),
+    StructField("client_id",             StringType(),  True),
+    StructField("country",               StringType(),  True),
+    StructField("default_search_engine", StringType(),  True),
+    StructField("distribution_id",       StringType(),  True),
+    StructField("locale",                StringType(),  True),
+    StructField("normalized_channel",    StringType(),  True),
+    StructField("profile_creation_date", LongType(),    True),
+    StructField("submission_date_s3",    StringType(),  False),
+    StructField("subsession_length",     LongType(),    True),
+    StructField("subsession_start_date", StringType(),  True),
+    StructField("sync_configured",       BooleanType(), True),
+    StructField("sync_count_desktop",    IntegerType(), True),
+    StructField("sync_count_mobile",     IntegerType(), True),
+    StructField("timestamp",             LongType(),    True),
+    StructField("total_uri_count",       IntegerType(), True),
+    StructField("unique_domains_count", IntegerType(), True)])
+
+default_sample = {
+    "app_version":           "57.0.0",
+    "attribution": {
+        "source": "source-value",
+        "medium": "medium-value",
+        "campaign": "campaign-value",
+        "content": "content-value"
+    },
+    "channel":               "release",
+    "client_id":             "client-id",
+    "country":               "US",
+    "default_search_engine": "wikipedia",
+    "distribution_id":       "mozilla42",
+    "locale":                "en-US",
+    "normalized_channel":    "release",
+    "profile_creation_date": 17181,
+    "submission_date_s3":    "20170115",
+    "subsession_length":     1000,
+    "subsession_start_date": "2017-01-15",
+    "sync_configured":       False,
+    "sync_count_desktop":    1,
+    "sync_count_mobile":     1,
+    "timestamp":             1491244610603260700,  # microseconds
```

Comment:
You're right, I trusted this [epoch converter](https://www.epochconverter.com/) as little too much...